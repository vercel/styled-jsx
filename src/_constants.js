export const STYLE_ATTRIBUTE = 'jsx'
export const GLOBAL_ATTRIBUTE = 'global'
export const STYLE_COMPONENT_ID = 'styleId'
export const MARKUP_ATTRIBUTE = 'data-jsx'
export const MARKUP_ATTRIBUTE_EXTERNAL = 'data-jsx-ext'
export const STYLE_COMPONENT = '_JSXStyle'
export const STYLE_COMPONENT_CSS = 'css'
