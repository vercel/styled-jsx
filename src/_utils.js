import * as t from 'babel-types'
import escapeStringRegExp from 'escape-string-regexp'
import traverse from 'babel-traverse'
import { parse } from 'babylon'
import { parse as parseCss } from 'css-tree'
import { SourceMapGenerator } from 'source-map'
import convert from 'convert-source-map'

import {
  STYLE_ATTRIBUTE,
  GLOBAL_ATTRIBUTE,
  STYLE_COMPONENT_ID,
  STYLE_COMPONENT,
  STYLE_COMPONENT_CSS
} from './_constants'

export const isGlobalEl = el =>
  el.attributes.some(({ name }) => name && name.name === GLOBAL_ATTRIBUTE)

export const isStyledJsx = ({ node: el }) =>
  t.isJSXElement(el) &&
  el.openingElement.name.name === 'style' &&
  el.openingElement.attributes.some(attr => attr.name.name === STYLE_ATTRIBUTE)

export const findStyles = path => {
  if (isStyledJsx(path)) {
    const { node } = path
    return isGlobalEl(node.openingElement) ? [path] : []
  }

  return path.get('children').filter(isStyledJsx)
}

export const getExpressionText = expr => {
  const node = expr.node

  // Assume string literal
  if (t.isStringLiteral(node)) {
    return node.value
  }

  const expressions = expr.get('expressions')

  // Simple template literal without expressions
  if (expressions.length === 0) {
    return node.quasis[0].value.cooked
  }

  // Special treatment for template literals that contain expressions:
  //
  // Expressions are replaced with a placeholder
  // so that the CSS compiler can parse and
  // transform the css source string
  // without having to know about js literal expressions.
  // Later expressions are restored
  // by doing a replacement on the transformed css string.
  //
  // e.g.
  // p { color: ${myConstant}; }
  // becomes
  // p { color: %%styled-jsx-placeholder-${id}%%; }

  const replacements = expressions
    .map((e, id) => ({
      pattern: new RegExp(
        `\\$\\{\\s*${escapeStringRegExp(e.getSource())}\\s*\\}`
      ),
      replacement: `%%styled-jsx-placeholder-${id}%%`,
      initial: `$\{${e.getSource()}}`
    }))
    .sort((a, b) => a.initial.length < b.initial.length)

  const source = expr.getSource().slice(1, -1)

  const modified = replacements.reduce((source, currentReplacement) => {
    source = source.replace(
      currentReplacement.pattern,
      currentReplacement.replacement
    )
    return source
  }, source)

  return {
    source,
    modified,
    replacements
  }
}

export const restoreExpressions = (css, replacements) =>
  replacements.reduce((css, currentReplacement) => {
    css = css.replace(
      new RegExp(currentReplacement.replacement, 'g'),
      currentReplacement.initial
    )
    return css
  }, css)

export const makeStyledJsxCss = (transformedCss, isTemplateLiteral) => {
  if (!isTemplateLiteral) {
    return t.stringLiteral(transformedCss)
  }
  // Build the expression from transformedCss
  let css
  traverse(parse(`\`${transformedCss}\``), {
    TemplateLiteral(path) {
      if (!css) {
        css = path.node
      }
    }
  })
  return css
}

export const makeStyledJsxTag = (id, transformedCss, isTemplateLiteral) => {
  let css

  if (
    typeof transformedCss === 'object' &&
    (t.isIdentifier(transformedCss) || t.isMemberExpression(transformedCss))
  ) {
    css = transformedCss
  } else {
    css = makeStyledJsxCss(transformedCss, isTemplateLiteral)
  }

  return t.jSXElement(
    t.jSXOpeningElement(
      t.jSXIdentifier(STYLE_COMPONENT),
      [
        t.jSXAttribute(
          t.jSXIdentifier(STYLE_COMPONENT_ID),
          t.jSXExpressionContainer(
            typeof id === 'number' ? t.numericLiteral(id) : id
          )
        ),
        t.jSXAttribute(
          t.jSXIdentifier(STYLE_COMPONENT_CSS),
          t.jSXExpressionContainer(css)
        )
      ],
      true
    ),
    null,
    []
  )
}

// We only allow constants to be used in template literals.
// The following visitor ensures that MemberExpressions and Identifiers
// are not in the scope of the current Method (render) or function (Component).
export const validateExpressionVisitor = {
  MemberExpression(path, scope) {
    const { node } = path
    if (
      (t.isIdentifier(node.property) &&
        (t.isThisExpression(node.object) &&
          (node.property.name === 'props' ||
            node.property.name === 'state'))) ||
      (t.isIdentifier(node.object) && scope.hasOwnBinding(node.object.name))
    ) {
      throw path.buildCodeFrameError(
        `Expected a constant ` +
          `as part of the template literal expression ` +
          `(eg: <style jsx>{\`p { color: $\{myColor}\`}</style>), ` +
          `but got a MemberExpression: this.${node.property.name}`
      )
    }
  },
  Identifier(path, scope) {
    if (t.isMemberExpression(path.parentPath)) {
      return
    }
    const { name } = path.node
    if (scope.hasOwnBinding(name)) {
      throw path.buildCodeFrameError(
        `Expected \`${name}\` ` +
          `to not come from the closest scope.\n` +
          `Styled JSX encourages the use of constants ` +
          `instead of \`props\` or dynamic values ` +
          `which are better set via inline styles or \`className\` toggling. ` +
          `See https://github.com/zeit/styled-jsx#dynamic-styles`
      )
    }
  }
}

export const validateExpression = (expr, scope) =>
  expr.traverse(validateExpressionVisitor, scope)

export const generateAttribute = (name, value) =>
  t.jSXAttribute(t.jSXIdentifier(name), t.jSXExpressionContainer(value))

export const isValidCss = str => {
  try {
    parseCss(
      // Replace the placeholders with some valid CSS
      // so that parsing doesn't fail for otherwise valid CSS.
      str
        // Replace all the placeholders with `all`
        .replace(
          // `\S` (the `delimiter`) is to match
          // the beginning of a block `{`
          // a property `:`
          // or the end of a property `;`
          /(\S)?\s*%%styled-jsx-placeholder-[^%]+%%(?:\s*(\}))?/gi,
          (match, delimiter, isBlockEnd) => {
            // The `end` of the replacement would be
            let end

            if (delimiter === ':' && isBlockEnd) {
              // ';}' single property block without semicolon
              // E.g. { color: all;}
              end = `;}`
            } else if (delimiter === '{' || isBlockEnd) {
              // ':;' when we are at the beginning or the end of a block
              // E.g. { all:; ...otherstuff
              // E.g. all:; }
              end = `:;${isBlockEnd || ''}`
            } else if (delimiter === ';') {
              // ':' when we are inside of a block
              // E.g. color: red; all:; display: block;
              end = ':'
            } else {
              // Otherwise empty
              end = ''
            }

            return `${delimiter || ''}all${end}`
          }
        )
        // Replace block placeholders before media queries
        // E.g. all @media (all) {}
        .replace(/all\s*([@])/g, (match, delimiter) => `all {} ${delimiter}`)
        // Replace block placeholders at the beginning of a media query block
        // E.g. @media (all) { all:; div { ... }}
        .replace(/@media[^{]+{\s*all:;/g, '@media (all) { ')
    )
    return true
  } catch (err) {}
  return false
}

export const makeSourceMapGenerator = file => {
  const filename = file.opts.sourceFileName
  const generator = new SourceMapGenerator({
    file: filename,
    sourceRoot: file.opts.sourceRoot
  })

  generator.setSourceContent(filename, file.code)
  return generator
}

export const addSourceMaps = (code, generator, filename) =>
  [
    code,
    convert.fromObject(generator).toComment({ multiline: true }),
    `/*@ sourceURL=${filename} */`
  ].join('\n')

export const combinePlugins = (plugins, opts) => {
  if (!plugins) {
    return css => css
  }

  if (
    !Array.isArray(plugins) ||
    plugins.some(p => !Array.isArray(p) && typeof p !== 'string')
  ) {
    throw new Error(
      '`plugins` must be an array of plugins names (string) or an array `[plugin-name, {options}]`'
    )
  }

  return plugins
    .map((plugin, i) => {
      let options = {}
      if (Array.isArray(plugin)) {
        options = plugin[1] || {}
        plugin = plugin[0]
      }

      // eslint-disable-next-line import/no-dynamic-require
      let p = require(plugin)
      if (p.default) {
        p = p.default
      }

      const type = typeof p
      if (type !== 'function') {
        throw new Error(
          `Expected plugin ${plugins[i]} to be a function but instead got ${type}`
        )
      }
      return {
        plugin: p,
        settings: {
          ...opts,
          options
        }
      }
    })
    .reduce(
      (previous, { plugin, settings }) => css =>
        plugin(previous ? previous(css) : css, settings),
      null
    )
}
