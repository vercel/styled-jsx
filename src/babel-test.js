import jsx from 'babel-plugin-syntax-jsx'

export default function() {
  return {
    inherits: jsx,
    visitor: {
      JSXOpeningElement(path) {
        const el = path.node
        const { name } = el.name || {}

        if (name !== 'style') {
          return
        }

        el.attributes = el.attributes.filter(a => a.name.name !== 'jsx')
      }
    }
  }
}
