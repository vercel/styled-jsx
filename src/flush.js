// Ours
import memory from './memory'

export default function flush() {
  const ret = {}

  for (const i in memory) {
    if (!{}.hasOwnProperty.call(memory, i)) {
      continue
    }

    ret[i] = memory[i]
    delete memory[i]
  }

  return ret
}
