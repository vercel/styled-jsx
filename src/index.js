export {
  StyleRegistry,
  createStyleRegistry,
  useStyleRegistry
} from './stylesheet-registry'
