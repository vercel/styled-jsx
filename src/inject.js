// Ours
import memory from './memory'

const isBrowser = typeof window !== 'undefined'
const tags = {}

export default function inject(id, css) {
  if (isBrowser) {
    // if the tag is already present we ignore it!
    if (!tags[id]) {
      const el = makeStyleTag(css)

      tags[id] = el
      memory[id] = el
    }
  } else {
    memory[id] = css
  }
}

function makeStyleTag(str) {
  // based on implementation by glamor
  const tag = document.createElement('style')

  tag.type = 'text/css'
  tag.appendChild(document.createTextNode(str))

  const head = document.head || document.getElementsByTagName('head')[0]
  head.appendChild(tag)

  return tag
}
