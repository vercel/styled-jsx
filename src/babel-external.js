import * as t from 'babel-types'

import { STYLE_COMPONENT } from './_constants'

import {
  getJSXStyleInfo,
  processCss,
  cssToBabelType,
  validateExternalExpressions,
  getScope,
  computeClassNames,
  makeStyledJsxTag,
  createReactComponentImportDeclaration,
  setStateOptions
} from './_utils'

const isModuleExports = t.buildMatchMemberExpression('module.exports')

export function processTaggedTemplateExpression({
  type,
  path,
  fileInfo,
  splitRules,
  plugins,
  vendorPrefix
}) {
  const templateLiteral = path.get('quasi')
  let scope

  // Check whether there are undefined references or
  // references to this.something (e.g. props or state).
  // We allow dynamic styles only when resolving styles.
  if (type !== 'resolve') {
    validateExternalExpressions(templateLiteral)
  } else if (!path.scope.path.isProgram()) {
    scope = getScope(path)
  }

  const stylesInfo = getJSXStyleInfo(templateLiteral, scope)

  const { staticClassName, className } = computeClassNames([stylesInfo])

  const styles = processCss(
    {
      ...stylesInfo,
      staticClassName,
      fileInfo,
      isGlobal: type === 'global',
      plugins,
      vendorPrefix
    },
    { splitRules }
  )

  if (type === 'resolve') {
    const { hash, css, expressions } = styles
    path.replaceWith(
      // {
      //   styles: <_JSXStyle ... />,
      //   className: 'jsx-123'
      // }
      t.objectExpression([
        t.objectProperty(
          t.identifier('styles'),
          makeStyledJsxTag(hash, css, expressions)
        ),
        t.objectProperty(t.identifier('className'), className)
      ])
    )
    return
  }

  const id = path.parentPath.node.id
  const baseExportName = id ? id.name : 'default'
  let parentPath =
    baseExportName === 'default'
      ? path.parentPath
      : path.findParent(
          path =>
            path.isVariableDeclaration() ||
            (path.isAssignmentExpression() &&
              isModuleExports(path.get('left').node))
        )

  if (baseExportName !== 'default' && !parentPath.parentPath.isProgram()) {
    parentPath = parentPath.parentPath
  }

  const css = cssToBabelType(styles.css)
  const newPath = t.isArrayExpression(css)
    ? css
    : t.newExpression(t.identifier('String'), [css])

  // default exports

  if (baseExportName === 'default') {
    const defaultExportIdentifier = path.scope.generateUidIdentifier(
      'defaultExport'
    )
    parentPath.insertBefore(
      t.variableDeclaration('const', [
        t.variableDeclarator(defaultExportIdentifier, newPath)
      ])
    )

    parentPath.insertBefore(addHash(defaultExportIdentifier, styles.hash))
    path.replaceWith(defaultExportIdentifier)
    return
  }

  // local and named exports

  parentPath.insertAfter(addHash(t.identifier(baseExportName), styles.hash))
  path.replaceWith(newPath)
}

function addHash(exportIdentifier, hash) {
  const value = typeof hash === 'string' ? t.stringLiteral(hash) : hash
  return t.expressionStatement(
    t.assignmentExpression(
      '=',
      t.memberExpression(exportIdentifier, t.identifier('__hash')),
      value
    )
  )
}

export const visitor = {
  ImportDeclaration(path, state) {
    // import css from 'styled-jsx/css'
    if (path.node.source.value !== 'styled-jsx/css') {
      return
    }

    // Find all the imported specifiers.
    // e.g import css, { global, resolve } from 'styled-jsx/css'
    // -> ['css', 'global', 'resolve']
    const specifiersNames = path.node.specifiers.map(
      specifier => specifier.local.name
    )
    specifiersNames.forEach(tagName => {
      // Get all the reference paths i.e. the places that use the tagName above
      // eg.
      // css`div { color: red }`
      // css.global`div { color: red }`
      // global`div { color: red `
      const binding = path.scope.getBinding(tagName)

      if (!binding || !Array.isArray(binding.referencePaths)) {
        return
      }

      // Produces an object containing all the TaggedTemplateExpression paths detected.
      // The object contains { scoped, global, resolve }
      const taggedTemplateExpressions = binding.referencePaths
        .map(ref => ref.parentPath)
        .reduce(
          (result, path) => {
            let taggedTemplateExpression
            if (path.isTaggedTemplateExpression()) {
              // css`` global`` resolve``
              taggedTemplateExpression = path
            } else if (
              path.parentPath &&
              path.isMemberExpression() &&
              path.parentPath.isTaggedTemplateExpression()
            ) {
              // This part is for css.global`` or css.resolve``
              // using the default import css
              taggedTemplateExpression = path.parentPath
            } else {
              return result
            }

            const tag = taggedTemplateExpression.get('tag')
            const id = tag.isIdentifier()
              ? tag.node.name
              : tag.get('property').node.name

            if (result[id]) {
              result[id].push(taggedTemplateExpression)
            } else {
              result.scoped.push(taggedTemplateExpression)
            }
            return result
          },
          {
            scoped: [],
            global: [],
            resolve: []
          }
        )

      let hasJSXStyle = false

      const { vendorPrefix, sourceMaps } = state.opts

      Object.keys(taggedTemplateExpressions).forEach(type =>
        taggedTemplateExpressions[type].forEach(path => {
          hasJSXStyle = true
          // Process each css block
          processTaggedTemplateExpression({
            type,
            path,
            fileInfo: {
              file: state.file,
              sourceFileName: state.file.opts.sourceFileName,
              sourceMaps
            },
            splitRules:
              typeof state.opts.optimizeForSpeed === 'boolean'
                ? state.opts.optimizeForSpeed
                : process.env.NODE_ENV === 'production',
            plugins: state.plugins,
            vendorPrefix
          })
        })
      )

      // When using the `resolve` helper we need to add an import
      // for the _JSXStyle component `styled-jsx/style`
      if (
        hasJSXStyle &&
        taggedTemplateExpressions.resolve.length > 0 &&
        !state.hasInjectedJSXStyle &&
        !path.scope.hasBinding(STYLE_COMPONENT)
      ) {
        state.hasInjectedJSXStyle = true
        const importDeclaration = createReactComponentImportDeclaration()
        path.scope.path.node.body.unshift(importDeclaration)
      }
    })

    // Finally remove the import
    path.remove()
  }
}

export default function() {
  return {
    Program(path, state) {
      setStateOptions(state)
    },
    ...visitor
  }
}
