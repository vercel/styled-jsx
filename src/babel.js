// Packages
import jsx from 'babel-plugin-syntax-jsx'

import { visitor as externalStylesVisitor } from './babel-external'

import {
  isGlobalEl,
  isStyledJsx,
  findStyles,
  makeStyledJsxTag,
  getJSXStyleInfo,
  computeClassNames,
  addClassName,
  getScope,
  processCss,
  createReactComponentImportDeclaration,
  setStateOptions
} from './_utils'

import { STYLE_COMPONENT } from './_constants'

export default function({ types: t }) {
  return {
    inherits: jsx,
    visitor: {
      JSXOpeningElement(path, state) {
        const el = path.node
        const { name } = el.name || {}

        if (!state.hasJSXStyle) {
          return
        }

        if (state.ignoreClosing === null) {
          // We keep a counter of elements inside so that we
          // can keep track of when we exit the parent to reset state
          // note: if we wished to add an option to turn off
          // selectors to reach parent elements, it would suffice to
          // set this to `1` and do an early return instead
          state.ignoreClosing = 0
        }

        if (
          name &&
          name !== 'style' &&
          name !== STYLE_COMPONENT &&
          name.charAt(0) !== name.charAt(0).toUpperCase()
        ) {
          if (state.className) {
            addClassName(path, state.className)
          }
        }

        state.ignoreClosing++
        // Next visit will be: JSXElement exit()
      },
      JSXElement: {
        enter(path, state) {
          if (state.hasJSXStyle !== null) {
            return
          }

          const styles = findStyles(path)

          if (styles.length === 0) {
            return
          }

          state.styles = []
          state.externalStyles = []

          const scope = getScope(path)

          for (const style of styles) {
            // Compute children excluding whitespace
            const children = style.get('children').filter(
              c =>
                t.isJSXExpressionContainer(c.node) ||
                // Ignore whitespace around the expression container
                (t.isJSXText(c.node) && c.node.value.trim() !== '')
            )

            if (children.length !== 1) {
              throw path.buildCodeFrameError(
                `Expected one child under ` +
                  `JSX Style tag, but got ${children.length} ` +
                  `(eg: <style jsx>{\`hi\`}</style>)`
              )
            }

            const child = children[0]

            if (!t.isJSXExpressionContainer(child)) {
              throw path.buildCodeFrameError(
                `Expected a child of ` +
                  `type JSXExpressionContainer under JSX Style tag ` +
                  `(eg: <style jsx>{\`hi\`}</style>), got ${child.type}`
              )
            }

            const expression = child.get('expression')

            if (t.isIdentifier(expression)) {
              const idName = expression.node.name
              if (expression.scope.hasBinding(idName)) {
                const externalStylesIdentifier = t.identifier(idName)
                const isGlobal = isGlobalEl(style.get('openingElement').node)
                state.externalStyles.push([
                  t.memberExpression(
                    externalStylesIdentifier,
                    t.identifier(isGlobal ? '__hash' : '__scopedHash')
                  ),
                  externalStylesIdentifier,
                  isGlobal
                ])
                continue
              }

              throw path.buildCodeFrameError(
                `The Identifier ` +
                  `\`${expression.getSource()}\` is either \`undefined\` or ` +
                  `it is not an external StyleSheet reference i.e. ` +
                  `it doesn't come from an \`import\` or \`require\` statement`
              )
            }

            if (
              !t.isTemplateLiteral(expression) &&
              !t.isStringLiteral(expression)
            ) {
              throw path.buildCodeFrameError(
                `Expected a template ` +
                  `literal or String literal as the child of the ` +
                  `JSX Style tag (eg: <style jsx>{\`some css\`}</style>),` +
                  ` but got ${expression.type}`
              )
            }

            state.styles.push(getJSXStyleInfo(expression, scope))
          }

          let externalJsxId
          if (state.externalStyles.length > 0) {
            const expressions = state.externalStyles
              // Remove globals
              .filter(s => !s[2])
              .map(s => s[0])

            const expressionsLength = expressions.length

            if (expressionsLength === 0) {
              externalJsxId = null
            } else {
              // Construct a template literal of this form:
              // `jsx-${styles.__scopedHash} jsx-${otherStyles.__scopedHash}`
              externalJsxId = t.templateLiteral(
                [
                  t.templateElement({ raw: 'jsx-', cooked: 'jsx-' }),
                  ...[...new Array(expressionsLength - 1)].map(() =>
                    t.templateElement({ raw: ' jsx-', cooked: ' jsx-' })
                  ),
                  t.templateElement({ raw: '', cooked: '' }, true)
                ],
                expressions
              )
            }
          }

          if (state.styles.length > 0 || externalJsxId) {
            const { staticClassName, className } = computeClassNames(
              state.styles,
              externalJsxId
            )
            state.className = className
            state.staticClassName = staticClassName
          }

          state.hasJSXStyle = true
          state.file.hasJSXStyle = true
          // Next visit will be: JSXOpeningElement
        },
        exit(path, state) {
          const isGlobal = isGlobalEl(path.node.openingElement)
          if (state.hasJSXStyle && !--state.ignoreClosing && !isGlobal) {
            state.hasJSXStyle = null
            state.className = null
            state.externalJsxId = null
          }

          if (!state.hasJSXStyle || !isStyledJsx(path)) {
            return
          }

          if (state.ignoreClosing > 1) {
            let styleTagSrc
            try {
              styleTagSrc = path.getSource()
            } catch (err) {}
            throw path.buildCodeFrameError(
              'Detected nested style tag' +
                (styleTagSrc ? `: \n\n${styleTagSrc}\n\n` : ' ') +
                'styled-jsx only allows style tags ' +
                'to be direct descendants (children) of the outermost ' +
                'JSX element i.e. the subtree root.'
            )
          }

          if (
            state.externalStyles.length > 0 &&
            path.get('children').filter(child => {
              if (!t.isJSXExpressionContainer(child)) {
                return false
              }
              const expression = child.get('expression')
              return expression && expression.isIdentifier()
            }).length === 1
          ) {
            const [id, css, isGlobal] = state.externalStyles.shift()

            path.replaceWith(
              makeStyledJsxTag(
                id,
                isGlobal
                  ? css
                  : t.memberExpression(
                      t.identifier(css.name),
                      t.identifier('__scoped')
                    )
              )
            )
            return
          }

          const { vendorPrefixes, sourceMaps } = state.opts
          const stylesInfo = {
            ...state.styles.shift(),
            fileInfo: {
              file: state.file,
              sourceFileName:
                state.file.opts.sourceFileName || state.file.sourceFileName,
              sourceMaps,
              filename: state.file.opts.filename || state.file.filename
            },
            staticClassName: state.staticClassName,
            isGlobal,
            plugins: state.plugins,
            vendorPrefixes
          }
          const splitRules =
            typeof state.opts.optimizeForSpeed === 'boolean'
              ? state.opts.optimizeForSpeed
              : process.env.NODE_ENV === 'production'

          const { hash, css, expressions } = processCss(stylesInfo, {
            splitRules
          })

          path.replaceWith(makeStyledJsxTag(hash, css, expressions))
        }
      },
      Program: {
        enter(path, state) {
          state.hasJSXStyle = null
          state.ignoreClosing = null
          state.file.hasJSXStyle = false

          setStateOptions(state)
        },
        exit({ node, scope }, state) {
          if (
            !(
              state.file.hasJSXStyle &&
              !state.hasInjectedJSXStyle &&
              !scope.hasBinding(STYLE_COMPONENT)
            )
          ) {
            return
          }
          state.hasInjectedJSXStyle = true
          const importDeclaration = createReactComponentImportDeclaration()
          node.body.unshift(importDeclaration)
        }
      },

      // Transpile external styles
      ...externalStylesVisitor
    }
  }
}
