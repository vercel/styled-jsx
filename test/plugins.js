// Packages
import test from 'ava'

// Ours
import babelPlugin from '../src/babel'
import { combinePlugins } from '../src/_utils'
import _transform from './_transform'
import testPlugin1 from './fixtures/plugins/plugin'
import testPlugin2 from './fixtures/plugins/another-plugin'

const transform = (file, opts = {}) =>
  _transform(file, {
    plugins: [
      [
        babelPlugin,
        { plugins: [require.resolve('./fixtures/plugins/another-plugin')] }
      ]
    ],
    ...opts
  })

test('combinePlugins returns an identity function when plugins is undefined', t => {
  const test = 'test'
  const plugins = combinePlugins()
  t.is(plugins(test), test)
})

test('combinePlugins throws if plugins is not an array', t => {
  t.throws(() => {
    combinePlugins(2)
  })
})

test('combinePlugins throws if plugins is not an array of strings', t => {
  t.throws(() => {
    combinePlugins(['test', 2])
  })
})

test('combinePlugins throws if loaded plugins are not functions', t => {
  t.throws(() => {
    combinePlugins([
      require.resolve('./fixtures/plugins/plugin'),
      require.resolve('./fixtures/plugins/invalid-plugin')
    ])
  })
})

test('combinePlugins works with a single plugin', t => {
  const plugins = combinePlugins([require.resolve('./fixtures/plugins/plugin')])

  t.is(testPlugin1('test'), plugins('test'))
})

test('combinePlugins works with options', t => {
  const expectedOption = 'my-test'
  const plugins = combinePlugins([
    [
      require.resolve('./fixtures/plugins/options'),
      {
        test: expectedOption
      }
    ]
  ])
  t.is(plugins(''), expectedOption)
})

test('combinePlugins applies plugins left to right', t => {
  const plugins = combinePlugins([
    require.resolve('./fixtures/plugins/plugin'),
    require.resolve('./fixtures/plugins/another-plugin')
  ])

  t.is(testPlugin2(testPlugin1('test')), plugins('test'))
})

test('applies plugins', async t => {
  const { code } = await transform('./fixtures/with-plugins.js')
  t.snapshot(code)
})
