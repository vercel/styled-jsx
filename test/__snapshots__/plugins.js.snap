// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applies plugins 1`] = `
"import _JSXStyle from 'styled-jsx/style';
import styles from './styles';
const color = 'red';

export default (() => <div data-jsx={4216192053} data-jsx-ext={styles.__scopedHash}>
    <p data-jsx={4216192053} data-jsx-ext={styles.__scopedHash}>test</p>
    <_JSXStyle styleId={4216192053} css={\`span.\${color}[data-jsx=\\"4216192053\\"]{color:\${otherColor}}\`} />
    <_JSXStyle styleId={styles.__scopedHash} css={styles.__scoped} />
  </div>);"
`;
